# 7. Aggregating Data

#### Importing libraries and dataframe
#### Aggregating data using transform()
#### Creating new column named 'loyalty_flag' using aggregation
#### Creating a new column named 'spender_type_flag' using aggregation
#### Creating a new column named 'customer_frequency_flag' using aggregation

Import Libraries and Dataframe

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\Neena Tilton\Dropbox\Projects\01_2020_InstacartBasket'

In [3]:
# import main dataframe that I exported at end of task 4.7
ords_prods_merge = pd.read_pickle(os.path.join(path, '02_Data', 'PreparedData', 'orders_products_merged.pkl'))

Aggregating data using transform(): 

In [20]:
# Creating a new column named 'max_order' which is filled with maximum orders made by each user.
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [22]:
# Removing default Jupyter display max so we can see top 100 rows.
pd.options.display.max_rows = None

In [23]:
ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   
30   2542086       52            11                   2                  9   
31    944694       52            12                   1                 16   
32   1268191       52            15                   2                 11   
33   1122089       52            17                   2                 10   
34   1498922       52            18                   2                 10   
35    180919       52            19                   3                  8   
36   2695875       52            21                   5                 10   
37   3205960       67             1                   3                 11   
38    970922       67             2                   1                 15   
39    596585       67             3                   1                 13   
40   2037087       67             4                   2                  9   
41   2474111       67             5                   3                 14   
42   1568481       67             6                   3                 12   
43   2048527       67             8                   2                 11   
44   2389031       67             9                   2                 14   
45   2378562       67            10                   1                 10   
46   1515698       67            11                   1                 11   
47   1928050       67            12                   2                 12   
48   1949274       67            13                   1                 10   
49    551869       67            14                   1                  9   
50    768572       67     

Creating new column named 'loyalty_flag' categorizing customers into 'loyal', 'regular', and 'new':

In [24]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [25]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [26]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [27]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

Printing parts of the dataframe to only see 'user_id', 'loyalty_flag', and 'order_number' columns:

In [28]:
ords_prods_merge[['user_id','loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

Aggregating the average orders made by 'department_id':

In [31]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

We want to investigate whether there's a difference between spending habits of the three types of customers, using the loyalty flag:

In [32]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

Loyal customers tend to purchase items with mean price of 10.39, whereas New customers tend to purchase items with mean price of 13.29 and Regular customers with price of 12.49. Loyal customers make more frequent orders with lower priced items than New and Regular customers. 

Creating a new column named 'spender_type_flag' based on the average price of items purchased by user. The criteria is as follows:
 If the mean of the prices of products purchased by a user < 10, flag as "Low spender"
 If the mean of the prices of products purchased by a user >= 10, flas as "High spender"

In [33]:
# Creating a new column 'mean_price_of_product' with value of mean of prices of products purchased by user
ords_prods_merge['mean_price_of_product'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [36]:
# Displaying minimized dataframe of columns to check if aggregation is accurate. 
ords_prods_merge[['user_id', 'prices', 'mean_price_of_product']].head(100)

user_id  prices  mean_price_of_product
0         1     9.0               6.367797
1         1     9.0               6.367797
2         1     9.0               6.367797
3         1     9.0               6.367797
4         1     9.0               6.367797
5         1     9.0               6.367797
6         1     9.0               6.367797
7         1     9.0               6.367797
8         1     9.0               6.367797
9         1     9.0               6.367797
10       15     9.0               3.980556
11       15     9.0               3.980556
12       15     9.0               3.980556
13       15     9.0               3.980556
14       15     9.0               3.980556
15       19     9.0               8.372059
16       19     9.0               8.372059
17       19     9.0               8.372059
18       21     9.0               7.012683
19       31     9.0               8.137793
20       31     9.0               8.137793
21       43     9.0               7.444898
22       43     9.0               7.444898
23       52     9.0               6.356287
24       52     9.0               6.356287
25       52     9.0               6.356287
26       52     9.0               6.356287
27       52     9.0               6.356287
28       52     9.0               6.356287
29       52     9.0               6.356287
30       52     9.0               6.356287
31       52     9.0               6.356287
32       52     9.0               6.356287
33       52     9.0               6.356287
34       52     9.0               6.356287
35       52     9.0               6.356287
36       52     9.0               6.356287
37       67     9.0               7.656790
38       67     9.0               7.656790
39       67     9.0               7.656790
40       67     9.0               7.656790
41       67     9.0               7.656790
42       67     9.0               7.656790
43       67     9.0               7.656790
44       67     9.0               7.656790
45       67     9.0               7.656790
46       67     9.0               7.656790
47       67     9.0               7.656790
48       67     9.0               7.656790
49       67     9.0               7.656790
50       67     9.0               7.656790
51       67     9.0               7.656790
52       67     9.0               7.656790
53       67     9.0               7.656790
54       67     9.0               7.656790
55       67     9.0               7.656790
56       81     9.0               8.165385
57       81     9.0               8.165385
58       82     9.0               6.441221
59       82     9.0               6.441221
60       82     9.0               6.441221
61       82     9.0               6.441221
62       82     9.0               6.441221
63       82     9.0               6.441221
64       82     9.0               6.441221
65       82     9.0               6.441221
66       82     9.0               6.441221
67       98     9.0               8.028000
68       98     9.0               8.028000
69       98     9.0               8.028000
70       98     9.0               8.028000
71       98     9.0               8.028000
72       98     9.0               8.028000
73       98     9.0               8.028000
74      109     9.0               5.850725
75      120     9.0               9.385714
76      120     9.0               9.385714
77      185     9.0               7.715000
78      195     9.0               7.149730
79      195     9.0               7.149730
80      195     9.0               7.149730
81      195     9.0               7.149730
82      195     9.0               7.149730
83      195     9.0               7.149730
84      195     9.0               7.149730
85      195     9.0               7.149730
86      195     9.0               7.149730
87      195     9.0               7.149730
88      195     9.0               7.149730
89      195     9.0               7.149730
90      222     9.0               6.862397
91      222     9.0               6.862397
92   

In [37]:
# Creating flag as "Low spender" or "High spender" using loc() function.
ords_prods_merge.loc[ords_prods_merge['mean_price_of_product'] < 10, 'spender_type_flag'] = 'Low spender'

In [38]:
ords_prods_merge.loc[ords_prods_merge['mean_price_of_product'] >= 10, 'spender_type_flag'] = 'High spender'

In [39]:
ords_prods_merge['spender_type_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spender_type_flag, dtype: int64

From this value count, we see that there are 634,245 spenders that are "High spenders".

Creating a new column named 'customer_frequency_flag' which categorizes the regularity of a user's ordering behavior according to the median in the "days_since_prior_order" column. Criteria is as follows: 
 If median of "days_since_prior_order" is > 20, flag as "Non-frequent customer"
 If median of "days_since_prior_order" is > 10 & <= 20, flag as "Regular customer"
 If median of "days_since_prior_order" is <= 10, flag as "Frequent customer"

In [41]:
ords_prods_merge['median_days_since_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [42]:
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] > 20, 'customer_frequency_flag'] = 'Non-frequent customer'

In [44]:
ords_prods_merge.loc[(ords_prods_merge['median_days_since_prior_order'] > 10) & (ords_prods_merge['median_days_since_prior_order'] <= 20), 'customer_frequency_flag'] = 'Regular customer'

In [45]:
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] <= 10, 'customer_frequency_flag'] = 'Frequent customer'

In [46]:
ords_prods_merge['customer_frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: customer_frequency_flag, dtype: int64

This analysis shows us that the majority of the Instacart customers are categorically "Frequent customers". In addition, we can also see that there were 5 users that made a single, first-time order, and never made an order again (see the row "NaN"). 

Exporting dataframe to pickle file:

In [48]:
ords_prods_merge.to_pickle(os.path.join(path, '02_Data', 'PreparedData', 'orders_products_merged.pkl'))